File: cane-cpp/cane-core/src/constraints/solvers.cpp
#include "solvers.hpp"

namespace cane::constraints {
float x_angle(const Affine& m) noexcept { return static_cast<float>(std::atan2(static_cast<double>(m.b), m.a)) * radians_to_degrees; }
float y_angle(const Affine& m) noexcept { return static_cast<float>(std::atan2(static_cast<double>(m.d), m.c)) * radians_to_degrees; }
float x_scale(const Affine& m) noexcept { return cane::hypot(m.a, m.b); }
float y_scale(const Affine& m) noexcept { return cane::hypot(m.c, m.d); }
float world_shear_y(const Affine& m) noexcept { return constraint_degrees(static_cast<float>(std::atan2(static_cast<double>(-m.c), m.d)) * radians_to_degrees - x_angle(m)); }
Affine rotate_axes(const Affine& m, float degrees) noexcept {
    const auto radians = degrees * degrees_to_radians;
    const auto co = static_cast<float>(std::cos(static_cast<double>(radians))), si = static_cast<float>(std::sin(static_cast<double>(radians)));
    return {co * m.a - si * m.b, si * m.a + co * m.b, co * m.c - si * m.d, si * m.c + co * m.d, m.tx, m.ty};
}
std::optional<BoneLocal> reconstruct_local(const BoneLocal& previous, const Affine& world, const Affine& parent, TransformMode mode) {
    const auto inverse = parent.inverse(matrix_epsilon); if (!inverse) return {};
    const auto position = inverse->transform({world.tx, world.ty}); Affine axes; float rotation_base = 0;
    if (mode == TransformMode::normal) axes = *inverse * world;
    else if (mode == TransformMode::only_translation) axes = world;
    else if (mode == TransformMode::no_rotation_or_reflection) {
        const auto basis = Affine::child(parent, {}, mode).inverse(matrix_epsilon); if (!basis) return {};
        axes = *basis * world;
    } else {
        const auto desired = x_angle(world); auto rotation = previous.rotation_degrees;
        for (unsigned i = 0; i < 12; ++i) {
            auto probe = previous; probe.rotation_degrees = rotation; const auto current = Affine::child(parent, probe, mode);
            if (x_scale(current) <= .00001f) return {};
            const auto error = constraint_degrees(desired - x_angle(current)); if (std::abs(error) <= .0001f) break;
            probe.rotation_degrees = rotation + .01f;
            const auto derivative = constraint_degrees(x_angle(Affine::child(parent, probe, mode)) - x_angle(current)) / .01f;
            if (std::abs(derivative) <= .00001f) return {};
            rotation += std::clamp(error / derivative, -45.0f, 45.0f);
        }
        rotation_base = rotation; BoneLocal base; base.rotation_degrees = rotation;
        const auto basis = Affine::child(parent, base, mode); const auto inverted = basis.inverse(matrix_epsilon);
        if (x_scale(basis) <= .00001f || !inverted) return {};
        axes = *inverted * world;
    }
    const auto sign_x = previous.scale_x < 0 ? -1.0f : 1.0f, sign_y = previous.scale_y < 0 ? -1.0f : 1.0f;
    const auto scale_x = x_scale(axes) * sign_x, scale_y = y_scale(axes) * sign_y;
    if (std::abs(scale_x) <= matrix_epsilon || std::abs(scale_y) <= matrix_epsilon) return {};
    const auto rotation = static_cast<float>(std::atan2(static_cast<double>(axes.b * sign_x), axes.a * sign_x)) * radians_to_degrees - previous.shear_x_degrees;
    const auto angle_y = static_cast<float>(std::atan2(static_cast<double>(-axes.c * sign_y), axes.d * sign_y)) * radians_to_degrees;
    const BoneLocal local{position.x, position.y, constraint_degrees(rotation_base + rotation), scale_x, scale_y, previous.shear_x_degrees, constraint_degrees(angle_y - rotation)};
    return Affine::from_local(local).finite() ? std::optional<BoneLocal>(local) : std::nullopt;
}
Affine Context::parent(std::size_t bone) const {
    const auto index = pose.data().bones().at(bone).parent_index; return index ? pose.world[*index] : root;
}
Affine Context::compose(std::size_t bone, const BoneLocal& local) const {
    return pose.data().bones()[bone].parent_index ? Affine::child(parent(bone), local, pose.modes[bone]) : root * Affine::from_local(local);
}
void Context::refresh_descendants(std::size_t bone) {
    auto& affected = scratch.affected; std::fill(affected.begin(), affected.end(), 0); affected.at(bone) = 1;
    const auto& bones = pose.data().bones();
    for (std::size_t i = bone; i < bones.size(); ++i) {
        if (i != bone && bones[i].parent_index && affected[*bones[i].parent_index]) {
            affected[i] = 1; pose.world[i] = Affine::child(pose.world[*bones[i].parent_index], pose.locals[i], pose.modes[i]);
        }
        if (affected[i] && !pose.world[i].finite()) throw Error(ErrorCode::non_finite, "apply", "Constraint hierarchy is non-finite.", "bones", bones[i].id);
    }
}
void Context::write_local(std::size_t bone, const BoneLocal& local) { pose.locals[bone] = local; pose.world[bone] = compose(bone, local); refresh_descendants(bone); }
void Context::set_world(std::size_t bone, const Affine& world) {
    if (!world.finite()) throw Error(ErrorCode::non_finite, "apply", "Constraint world matrix is non-finite.", "bones", pose.data().bones()[bone].id);
    pose.world[bone] = world;
    const auto local = reconstruct_local(pose.locals[bone], world, parent(bone), pose.data().bones()[bone].parent_index ? pose.modes[bone] : TransformMode::normal);
    if (local) pose.locals[bone] = *local;
}
void Context::write_world(std::size_t bone, const Affine& world) { set_world(bone, world); refresh_descendants(bone); }
}

File: cane-csharp/cane-core/Runtime/Constraints/Solvers.cs
using System;
using Cane.Animation;
using Cane.Format;

namespace Cane.Constraints
{
    internal static class Solvers
    {
        internal static float Degrees(float value) { float r = value % 360; if (r < 0) r += 360; return r > 180 ? r - 360 : r; }
        internal static float Positive(float value) { float r = value % 360; return r < 0 ? r + 360 : r; }
        internal static float Angle(Affine m) => Numeric.Atan2(m.B, m.A) * Numeric.RadToDeg;
        internal static float YAngle(Affine m) => Numeric.Atan2(m.D, m.C) * Numeric.RadToDeg;
        internal static float Sx(Affine m) => Numeric.Hypot(m.A, m.B);
        internal static float Sy(Affine m) => Numeric.Hypot(m.C, m.D);
        internal static float ShearY(Affine m) => Degrees(Numeric.Atan2(-m.C, m.D) * Numeric.RadToDeg - Angle(m));
        internal static void IncludeMax(ref float? maximum, float value) => maximum = maximum.HasValue ? Math.Max(maximum.Value, value) : value;
        internal static bool Inverse(Affine m, out Affine inverse)
        { if (Math.Abs(m.Determinant) <= Numeric.Epsilon) { inverse = default; return false; } return m.TryInverse(out inverse); }
        internal static Affine Parent(Pose pose, int bone, Affine root) => pose.Data.BoneData[bone].Parent < 0 ? root : pose.World[pose.Data.BoneData[bone].Parent];
        internal static void Solve(Pose pose, Affine root, SamplingOptions sampling, PhysicsStore physics, float delta, PhysicsEnvironment environment, float rootScaleX, float rootScaleY, PhysicsBudget budget)
        {
            foreach (Json constraint in pose.Constraints)
            {
                if (!pose.IsActive(constraint.S("id"), "constraintIds")) continue;
                string? target = constraint["targetBoneId"].StringOrNull;
                if (target != null && !pose.IsActive(target, "boneIds")) continue;
                switch (constraint.S("type"))
                {
                    case "ik": if (constraint.F("mix", 1) != 0) Ik.Solve(pose, constraint, root); break;
                    case "transform": Transform.Solve(pose, constraint, root); break;
                    case "path":
                        if (constraint.F("mixRotate", 1) != 0 || constraint.F("mixX", 1) != 0 || constraint.F("mixY", 1) != 0) PathConstraint.Solve(pose, constraint, root);
                        break;
                    case "physics": Physics.Solve(pose, constraint, root, physics, delta, environment, rootScaleX, rootScaleY, budget); break;
                    case "slider": if (constraint.F("mix", 1) != 0) Slider.Solve(pose, Pose.CopyConstraintState(constraint), root, sampling); break;
                }
            }
        }
        internal static void WriteLocal(Pose pose, int bone, BoneLocal local, Affine root)
        {
            pose.Locals[bone] = local; int parent = pose.Data.BoneData[bone].Parent;
            pose.World[bone] = parent < 0 ? root * Affine.FromLocal(local) : Affine.Child(pose.World[parent], local, pose.Modes[bone]);
            RefreshDescendants(pose, bone);
        }
        internal static void WriteWorld(Pose pose, int bone, Affine matrix, Affine root)
        { SetWorld(pose, bone, matrix, root); RefreshDescendants(pose, bone); }
        internal static void SetWorld(Pose pose, int bone, Affine matrix, Affine root)
        {
            if (!matrix.IsFinite) throw new RuntimeException(RuntimeErrorCode.NonFinite, "apply", "Constraint produced a non-finite matrix.", "bones", pose.Data.BoneData[bone].Id);
            pose.World[bone] = matrix;
            if (Reconstruct(pose.Locals[bone], matrix, Parent(pose, bone, root), pose.Data.BoneData[bone].Parent < 0 ? "normal" : pose.Modes[bone], out BoneLocal local)) pose.Locals[bone] = local;
        }
        internal static void RefreshDescendants(Pose pose, int start)
        {
            var changed = new bool[pose.World.Length]; changed[start] = true;
            for (int i = start + 1; i < changed.Length; i++)
            {
                int parent = pose.Data.BoneData[i].Parent;
                if (parent >= 0 && changed[parent]) { changed[i] = true; pose.World[i] = Affine.Child(pose.World[parent], pose.Locals[i], pose.Modes[i]); }
            }
            for (int i = start; i < changed.Length; i++) if (changed[i] && !pose.World[i].IsFinite)
                throw new RuntimeException(RuntimeErrorCode.NonFinite, "apply", "Constraint hierarchy overflowed.", "bones", pose.Data.BoneData[i].Id);
        }
        private static bool Reconstruct(BoneLocal previous, Affine world, Affine parent, string mode, out BoneLocal local)
        {
            local = previous; if (!Inverse(parent, out Affine inverse)) return false;
            Point position = inverse.Transform(world.Tx, world.Ty); Affine axes;
            float rotationBase = 0;
            if (mode == "normal") axes = inverse * world;
            else if (mode == "onlyTranslation") axes = world;
            else if (mode == "noRotationOrReflection")
            {
                Affine basis = Affine.Child(parent, BoneLocal.Identity, mode);
                if (!Inverse(basis, out Affine invBasis)) return false; axes = invBasis * world;
            }
            else
            {
                // Only no-scale inheritance depends nonlinearly on the local rotation.
                float desired = Angle(world), rotation = previous.RotationDegrees;
                for (int i = 0; i < 12; i++)
                {
                    Affine current = Affine.Child(parent, previous.With(2, rotation), mode);
                    if (Sx(current) <= 0.00001f) return false;
                    float error = Degrees(desired - Angle(current)); if (Math.Abs(error) <= 0.0001f) break;
                    Affine perturbed = Affine.Child(parent, previous.With(2, rotation + 0.01f), mode);
                    float derivative = Degrees(Angle(perturbed) - Angle(current)) / 0.01f;
                    if (Math.Abs(derivative) <= 0.00001f) return false;
                    rotation += Numeric.Clamp(error / derivative, -45, 45);
                }
                rotationBase = rotation;
                Affine basis = Affine.Child(parent, new BoneLocal(0, 0, rotation), mode);
                if (Sx(basis) <= 0.00001f || !Inverse(basis, out Affine invBasis)) return false;
                axes = invBasis * world;
            }
            float signX = previous.ScaleX < 0 ? -1 : 1, signY = previous.ScaleY < 0 ? -1 : 1;
            float scaleX = Sx(axes) * signX, scaleY = Sy(axes) * signY;
            if (Math.Abs(scaleX) <= Numeric.Epsilon || Math.Abs(scaleY) <= Numeric.Epsilon) return false;
            float rotationLocal = Numeric.Atan2(axes.B * signX, axes.A * signX) * Numeric.RadToDeg - previous.ShearXDegrees;
            float angleY = Numeric.Atan2(-axes.C * signY, axes.D * signY) * Numeric.RadToDeg;
            local = new BoneLocal(position.X, position.Y, Degrees(rotationBase + rotationLocal), scaleX, scaleY, previous.ShearXDegrees, Degrees(angleY - rotationLocal));
            return Affine.FromLocal(local).IsFinite;
        }
        internal static Affine Rotate(Affine m, float angle)
        {
            float co = Numeric.Cos(angle * Numeric.DegToRad), si = Numeric.Sin(angle * Numeric.DegToRad);
            return new Affine(co * m.A - si * m.B, si * m.A + co * m.B, co * m.C - si * m.D, si * m.C + co * m.D, m.Tx, m.Ty);
        }
    }
}
